pallets/file-system/src/utils.rs

```diff
@@ -578,96 +625,381 @@ where
-        let sp_id =
+        file_key_responses_input: FileKeyResponsesInput<T>,
+    ) -> Result<MspRespondStorageRequestsResult<T>, DispatchError> {
+        // Check that the sender is a Storage Provider and get its MSP ID
+        let msp_id =
             <T::Providers as shp_traits::ReadProvidersInterface>::get_provider_id(sender.clone())
                 .ok_or(Error::<T>::NotASp)?;
 
-        // Check that the sender is a MSP.
+        // Check that the sender is an MSP
         ensure!(
-            <T::Providers as ReadStorageProvidersInterface>::is_msp(&sp_id),
+            <T::Providers as ReadStorageProvidersInterface>::is_msp(&msp_id),
             Error::<T>::NotAMsp
         );
 
-        // Check that the sender corresponds to the MSP in the storage request and that it hasn't yet confirmed storing the file.
-        let (request_msp_id, confirm_status) = storage_request_metadata
-            .msp
-            .clone()
-            .ok_or(Error::<T>::RequestWithoutMsp)?;
-        ensure!(request_msp_id == sp_id, Error::<T>::NotSelectedMsp);
-        ensure!(confirm_status == false, Error::<T>::MspAlreadyConfirmed);
+        // Initialize batch responses
+        let mut batch_responses: BoundedVec<
+            BatchResponses<T>,
+            MaxBatchMspRespondStorageRequests<T>,
+        > = BoundedVec::default();
 
-        // Get the bucket ID from the storage request metadata
-        let bucket_id = storage_request_metadata.bucket_id;
+        // Preliminary check to ensure that the MSP is the one storing each bucket in the responses
+        for (bucket_id, _) in file_key_responses_input.iter() {
+            ensure!(
+                <T::Providers as ReadBucketsInterface>::is_bucket_stored_by_msp(
+                    &msp_id, &bucket_id
+                ),
+                Error::<T>::MspNotStoringBucket
+            );
+        }
 
-        // Check that the MSP is the one storing the bucket.
-        ensure!(
-            <T::Providers as ReadBucketsInterface>::is_bucket_stored_by_msp(&sp_id, &bucket_id),
-            Error::<T>::MspNotStoringBucket
-        );
+        // Process each bucket's responses
+        for (bucket_id, file_key_responses) in file_key_responses_input {
+            let mut failed: BoundedVec<
+                (MerkleHash<T>, DispatchError),
+                MaxBatchMspRespondStorageRequests<T>,
+            > = BoundedVec::default();
+
+            let owner = <T::Providers as ReadBucketsInterface>::get_bucket_owner(&bucket_id)
+                .map_err(|_| Error::<T>::BucketNotFound)?;
+
+            if let Some(accepted_file_keys) = file_key_responses.accept {
+                // Call do_msp_accept_storage_request, which returns the new_bucket_root
+                let (new_bucket_root, accepted_file_keys, failed_file_keys) =
+                    Self::do_msp_accept_storage_request(
+                        msp_id.clone(),
+                        bucket_id.clone(),
```

Comment:
Clippy warns about these two since they implement `Copy`, so no need to clone them :)